lib/ansible/modules/windows/win_domain_user.ps1

```diff
@@ -104,12 +207,32 @@ If ($state -eq 'present') {
     }
 
     # Set the password if required
-    If ($password -and (($new_user -and $update_password -eq "on_create") -or $update_password -eq "always")) {
-        $secure_password = ConvertTo-SecureString $password -AsPlainText -Force
-        Set-ADAccountPassword -Identity $username -Reset:$true -Confirm:$false -NewPassword $secure_password -WhatIf:$check_mode @extra_args
-        $user_obj = Get-ADUser -Identity $username -Properties * @extra_args
-        $result.password_updated = $true
-        $result.changed = $true
+    If (
+        (
+            $password
+        ) -and (
+            (
+                $new_user -and $update_password -eq "on_create"
+            ) -or (
+                $update_password -in @("always"; "when_changed")
+            )
+        )
+    ) {
+        If (-not $new_user) { # Don't uncecessary check if the credentials works, if we know it doesn't.
+            $test_new_credentials = Test-Credential -Username $username -Password $password
```

Comment:
We also don't want to test the credentials if `update_password` is `on_create` and we didn't create a new user. Right now this will always check and update the password for both `on_create` and `when_changed` effectively making them the same thing. We could always make this more efficient by not testing the credentials when `update_password` is `always` as we will always update it regardless of this result.

Reply:
No it won't $new_user is only true if the user was newly created https://github.com/ansible/ansible/pull/58383/files#diff-a814da53b65fce105cb5fb909b79b267R201 therefore it won't be touched.

The outer if-statement checks:
```
($password) ∧ (($new_user ∧ $update_password == "on_create") ∨ ($update_password ∈ {'always', 'when_chantged'}))
```

The inner if-statement:
```
¬($new_user)
```

So Test-Credential is executed when:
```
($password) ∧ (($new_user ∧ $update_password == "on_create") ∨ ($update_password ∈ {'always', 'when_chantged'})) ∧ ¬($new_user)
```

And for the case where `$password = 'something' ∧ update_password == 'on_create' ∧ $new_user == $true` (we just created the user) it will results in:

1. `$password` => $true
2. `(($new_user ∧ $update_password == "on_create")` => $true
3. ¬($new_user) => $false

=> So it will not be executed.

And for the case where `$password = 'something' ∧ update_password == 'on_create' ∧ $new_user == $false` (we the user already existed) it will results in:

1. `$password` => $true
2. `($new_user ∧ $update_password == "on_create")` => $false
3. `($update_password ∈ {'always', 'when_chantged'})` => $false

=> So it will not be executed either.

Reply:
Ahh yes I see it now sorry, we should still avoid testing the credentials when the credentials is `always` as we will be changing it anyway.

It did seem a bit confusing when I was reading through it originally so if you can simplify it then by all means go ahead.

Reply:
Sometimes you can't see the forest for the trees...
It's now way easier.